bolt/lib/Passes/ProfileQualityStats.cpp

```diff
@@ -0,0 +1,579 @@
+          continue;
+        }
+        TotalOutgoing += Count;
+        IncomingMap[Succ->getIndex()] += Count;
+        ++SuccBIIter;
+      }
+      OutgoingMap[BB.getIndex()] = TotalOutgoing;
+    }
+  }
+
+  // Initialize TotalMaxCountMaps and TotalMinCountMaps using
+  // TotalIncomingMaps and TotalOutgoingMaps
+  TotalFlowMapTy &TotalMaxCountMaps = TotalFlowMap.TotalMaxCountMaps;
+  TotalFlowMapTy &TotalMinCountMaps = TotalFlowMap.TotalMinCountMaps;
+  for (const auto &BFI : BC.getBinaryFunctions()) {
+    const BinaryFunction *Function = &BFI.second;
+    if (Function->empty() || !Function->hasValidProfile())
+      continue;
+    uint64_t FunctionNum = Function->getFunctionNumber();
+    FlowMapTy &IncomingMap = TotalIncomingMaps[FunctionNum];
+    FlowMapTy &OutgoingMap = TotalOutgoingMaps[FunctionNum];
+    FlowMapTy &MaxCountMap = TotalMaxCountMaps[FunctionNum];
+    FlowMapTy &MinCountMap = TotalMinCountMaps[FunctionNum];
+    for (const BinaryBasicBlock &BB : *Function) {
+      uint64_t BBNum = BB.getIndex();
+      MaxCountMap[BBNum] = std::max(IncomingMap[BBNum], OutgoingMap[BBNum]);
+      MinCountMap[BBNum] = std::min(IncomingMap[BBNum], OutgoingMap[BBNum]);
+    }
+  }
+
+  // Modify TotalMaxCountMaps and TotalMinCountMaps using call counts and
+  // fill out CallGraphIncomingMap
+  FunctionFlowMapTy &CallGraphIncomingMap = TotalFlowMap.CallGraphIncomingMap;
+  for (const auto &BFI : BC.getBinaryFunctions()) {
+    const BinaryFunction *Function = &BFI.second;
+    uint64_t FunctionNum = Function->getFunctionNumber();
+    FlowMapTy &MaxCountMap = TotalMaxCountMaps[FunctionNum];
+    FlowMapTy &MinCountMap = TotalMinCountMaps[FunctionNum];
+
+    // Update MaxCountMap, MinCountMap, and CallGraphIncomingMap
+    auto recordCall = [&](const BinaryBasicBlock *SourceBB,
+                          const MCSymbol *DestSymbol, uint64_t Count) {
+      if (Count == BinaryBasicBlock::COUNT_NO_PROFILE)
+        Count = 0;
+      const BinaryFunction *DstFunc =
+          DestSymbol ? BC.getFunctionForSymbol(DestSymbol) : nullptr;
+      if (DstFunc)
+        CallGraphIncomingMap[DstFunc->getFunctionNumber()] += Count;
+      if (SourceBB) {
+        unsigned BlockIndex = SourceBB->getIndex();
+        MaxCountMap[BlockIndex] = std::max(MaxCountMap[BlockIndex], Count);
+        MinCountMap[BlockIndex] = std::min(MinCountMap[BlockIndex], Count);
+      }
+    };
+
+    // Get pairs of (symbol, count) for each target at this callsite.
+    // If the call is to an unknown function the symbol will be nullptr.
+    // If there is no profiling data the count will be COUNT_NO_PROFILE.
+    using TargetDesc = std::pair<const MCSymbol *, uint64_t>;
+    using CallInfoTy = std::vector<TargetDesc>;
+    auto getCallInfo = [&](const BinaryBasicBlock *BB, const MCInst &Inst) {
+      CallInfoTy Counts;
+      const MCSymbol *DstSym = BC.MIB->getTargetSymbol(Inst);
+
+      // If this is an indirect call use perf data directly.
```

Comment:
I think it's a bit misleading to say we use perf data directly as we use attached annotation. I would drop the comment.